### PR TITLE
text to numeric encoding in to_... methods and some bug fixes

### LINKED ISSUE #14: Extract Dataset.Text strings to numpy
> Difficulty: ★★☆☆☆

## Background
`Dataset.Tabular` and `Dataset.Image` both have the methods `to_numpy()` and `to_pandas()`. 

`Dataset.Image` is read into memory using the Pillow library. Images can also be read either `to_pillow()` or `to_numpy()`. `Dataset.Image` disallows the creation of an `Encoderset` because it will use transforms/augmentation eventually.

NLP encodes either at the word "Bag of words" or character level. 

## Problem
We don't have a way to prepare raw text data as an array for a neural network. 

## Solution
- `to_numpy()` method that uses `sklearn.feature_extraction.text` methods like [`CountVectorizer ` and `TfidfVectorizer()`](https://scikit-learn.org/stable/modules/generated/sklearn.feature_extraction.text.TfidfTransformer.html#sklearn.feature_extraction.text.TfidfTransformer)
- `frequency:bool=True` argument in case they just want to return the counts.

## Discussion
- Alternatively, SpaCy has a Vocab object and seems much more language aware. However, it also seems a lot more involved, so I thought that this would be an easy way to get started.

### PR DESCRIPTION
**Pull request by sahilgupta2105**
Closes #14 #36 

Also, cleans up some validation logic while ingesting text data from the user. With this update, the user will be required to provide a 'TextData' column containing the text dataset if they use from_pandas(). This decision simplifies the downstream conversion of text data into numeric form as the feature columns are marked later on in the pipeline 

---

> *Put an `x` inside the applicable checkboxes `[ ]`.*

## Primary type of change:
- [ x] New feature.
- [ x] Bug fix.
- [ x] Refactor.
- [ ] Documentation.
- [ ] Build/ devops.
- [ ] Test.
- [ ] Other.

## Checklist:
- [ x] I have actually ran this code locally to make sure it works.
- [ x] I have pulled and merged the latest `main` into my feature branch.
- [ x] I have ran the existing [tests](https://github.com/aiqc/aiqc/new/main/.github#how-to-run-tests), and inspected areas that my code touches.
- [ ] I have updated the tests to include my changes.
- [ ] I have included updates to the documentation.
- [ ] I have built the documentation files `make html`.
- [ ] I am pushing a *branch*, not my *main/ master*.

> Describe the env you tested on: Mac OS 11, Python 3.7

## Contains breaking changes:
- [ ] Yes (explain what functionality it breaks).
- [x ] No

### CODE REVIEW

File: aiqc/__init__.py [hunk 1]
```diff
@@ -985,7 +1008,28 @@ def to_pandas(
...
+
+		
+		def to_numpy(
+			id:int, 
+			columns:list = None, 
+			samples:list = None
+		):
+			df = Dataset.Tabular.to_pandas(id, columns, samples)
+			word_counts, feature_names = Dataset.Text.get_feature_matrix(df)
+			return word_counts, feature_names
```
Comment (aiqc) [inline aiqc/__init__.py:1024]: Where is the numpy array 🙃 ?

image

sparse matrices do not work with Keras

File: aiqc/__init__.py [hunk 2]
```diff
@@ -985,7 +1008,28 @@ def to_pandas(
...
+
+
+		def get_feature_matrix(
+			dataframe:object
+		):
+			count_vect = CountVectorizer()
+			word_counts = count_vect.fit_transform(dataframe[Dataset.Text.column_name].tolist())
+			return word_counts, count_vect.get_feature_names()
```
Comment (aiqc) [inline aiqc/__init__.py:1032]: Keeping in mind for longterm, this metadata may be good for a Text table. We'll see.

File: aiqc/__init__.py [hunk 3]
```diff
@@ -985,7 +1008,28 @@ def to_pandas(
...
+			df = Dataset.Tabular.to_pandas(id, columns, samples)
+			word_counts, feature_names = Dataset.Text.get_feature_matrix(df)
+			return word_counts, feature_names
+
+
+		def get_feature_matrix(
+			dataframe:object
+		):
+			count_vect = CountVectorizer()
+			word_counts = count_vect.fit_transform(dataframe[Dataset.Text.column_name].tolist())
```
Comment (aiqc) [inline aiqc/__init__.py:1031]: if you `fit` then `transform` separately, then you can later use that `fit` object to `inverse_transform` the numbers back to text. Useful when making predictions and for human review.

These are actually pretty similar to the sklearn.preprocessing methods, so maybe they can go in Encoderset in the future. But keep charging down this path for now!

**Review by aiqc — APPROVED** [review]

File: aiqc/__init__.py (hunk 1 above)
Comment (sahilgupta2105) [reply aiqc/__init__.py:1024]: need to do more research on this then, I was skeptical about the to_pandas() method too, with large datasets a numpy array was would be highly inefficient and probably even not fit in memory, will be looping back to this question as I work through the other issues on writing tests for the NLP module

File: aiqc/__init__.py (hunk 2 above)
Comment (sahilgupta2105) [reply aiqc/__init__.py:1032]: okay, this is just boiler plate code to get started, planning on adding support for other transformation techniques as well, plus refactoring any meta data into tables as use cases show up in future

File: aiqc/__init__.py (hunk 3 above)
Comment (sahilgupta2105) [reply aiqc/__init__.py:1031]: so serialize the fit object and persist it in the database?

I agree, encoderset also seems like a good place (probably better) for this transformation as nothing useful is happening between dataset step and encoderset step by performing this transformation

**Pull request merged**